File: 0472-concatenated-words/0472-concatenated-words.py
class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        def dp(word):
            if len(word) == 0:
                return True
            
            for i in range(len(word)+1):
                if word[:i] in seen and dp(word[i:]):
                    return True
            return False
        ans = []
        seen = set(words)
        for word in words:
            seen.remove(word)

            if(dp(word)):
                ans.append(word)
            seen.add(word)
        return ans

File: 0472-concatenated-words/0472-concatenated-words.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
var findAllConcatenatedWordsInADict = function(words) {
    const seen = new Set(words)
    const dfs = (word)=>{
        for(let i = 0; i < word.length; i++)
            if (seen.has(word.slice(0,i+1)) && dfs(word.slice(i+1)))
                return true
        return word.length === 0
    
    }
    const ans = []
    for(let word of words){
        seen.delete(word)
        if(dfs(word))
            ans.push(word)
        seen.add(word)
    }
    return ans
};